File: sdk/typescript/rooch-sdk/src/keypairs/ed25519/publickey.ts
// Copyright (c) RoochNetwork
// SPDX-License-Identifier: Apache-2.0

import { fromB64 } from '@mysten/bcs'
import nacl from 'tweetnacl'

import { ROOCH_ADDRESS_LENGTH, RoochAddress } from '@/address'
import type { PublicKeyInitData } from '@/crypto'
import { PublicKey, SIGNATURE_SCHEME_TO_FLAG } from '@/crypto'
import { blake2b } from '@/utils'

const PUBLIC_KEY_SIZE = 32

/**
 * An Ed25519 public key
 */
export class Ed25519PublicKey extends PublicKey<RoochAddress> {
  static SIZE = PUBLIC_KEY_SIZE

  private readonly data: Uint8Array

  /**
   * Create a new Ed25519PublicKey object
   * @param value ed25519 public key as buffer or base-64 encoded string
   */
  constructor(value: PublicKeyInitData) {
    super()

    if (typeof value === 'string') {
      this.data = fromB64(value)
    } else if (value instanceof Uint8Array) {
      this.data = value
    } else {
      this.data = Uint8Array.from(value)
    }

    if (this.data.length !== PUBLIC_KEY_SIZE) {
      throw new Error(
        `Invalid public key input. Expected ${PUBLIC_KEY_SIZE} bytes, got ${this.data.length}`,
      )
    }
  }

  /**
   * Checks if two Ed25519 public keys are equal
   */
  override equals(publicKey: Ed25519PublicKey): boolean {
    return super.equals(publicKey)
  }

  /**
   * Return the byte array representation of the Ed25519 public key
   */
  toBytes(): Uint8Array {
    return this.data
  }

  /**
   * Return the Rooch address associated with this Ed25519 public key
   */
  flag(): number {
    return SIGNATURE_SCHEME_TO_FLAG.ED25519
  }

  /**
   * Verifies that the signature is valid for the provided message
   */
  async verify(message: Uint8Array, signature: Uint8Array): Promise<boolean> {
    return nacl.sign.detached.verify(message, signature, this.toBytes())
  }

  /**
   * Return the Rooch address associated with this Ed25519 public key
   */
  toAddress(): RoochAddress {
    const tmp = new Uint8Array(PUBLIC_KEY_SIZE + 1)
    tmp.set([SIGNATURE_SCHEME_TO_FLAG.ED25519])
    tmp.set(this.toBytes(), 1)

    // Each hex char represents half a byte, hence hex address doubles the length
    return new RoochAddress(blake2b(tmp, { dkLen: 32 }).slice(0, ROOCH_ADDRESS_LENGTH * 2))
  }
}

File: sdk/golang/rooch-sdk/keypairs/ed25519/publickey.go
package ed25519

import (
	"crypto/ed25519"
	"encoding/base64"
	"errors"
	"github.com/rooch-network/rooch/sdk/golang/rooch-sdk/address"
	"github.com/rooch-network/rooch/sdk/golang/rooch-sdk/crypto"
	"github.com/rooch-network/rooch/sdk/golang/rooch-sdk/utils"
)

const PUBLIC_KEY_SIZE = 32

// Ed25519PublicKey represents an Ed25519 public key
type Ed25519PublicKey struct {
	data []byte
}

// NewEd25519PublicKey creates a new Ed25519PublicKey object
// value can be either a base64 encoded string or a byte slice
func NewEd25519PublicKey(value interface{}) (*Ed25519PublicKey, error) {
	var data []byte

	switch v := value.(type) {
	case string:
		var err error
		data, err = base64.StdEncoding.DecodeString(v)
		if err != nil {
			return nil, err
		}
	case []byte:
		data = v
	default:
		return nil, errors.New("invalid public key input type")
	}

	if len(data) != PUBLIC_KEY_SIZE {
		return nil, errors.New("invalid public key size")
	}

	return &Ed25519PublicKey{
		data: data,
	}, nil
}

// Equals checks if two Ed25519 public keys are equal
func (pk *Ed25519PublicKey) Equals(other *Ed25519PublicKey) bool {
	if pk == nil || other == nil {
		return false
	}
	return crypto.BytesEqual(pk.data, other.data)
}

// ToBytes returns the byte array representation of the Ed25519 public key
func (pk *Ed25519PublicKey) ToBytes() []byte {
	return pk.data
}

// Flag returns the signature scheme flag for Ed25519
func (pk *Ed25519PublicKey) Flag() uint8 {
	return crypto.ED25519_FLAG
}

// Verify verifies that the signature is valid for the provided message
func (pk *Ed25519PublicKey) Verify(message, signature []byte) bool {
	return ed25519.Verify(pk.data, message, signature)
}

// ToAddress returns the Rooch address associated with this Ed25519 public key
func (pk *Ed25519PublicKey) ToAddress() *address.RoochAddress {
	tmp := make([]byte, PUBLIC_KEY_SIZE+1)
	tmp[0] = crypto.ED25519_FLAG
	copy(tmp[1:], pk.data)

	hash := utils.Blake2b(tmp, 32)
	addressBytes := hash[:address.ROOCH_ADDRESS_LENGTH*2]
	return address.NewRoochAddress(addressBytes)
} 